contracts/rust/deployer/src/lib.rs

```diff
@@ -1114,6 +1342,170 @@ mod tests {
+        )
+        .await?;
+        if upgrade_twice {
+            // upgrade to v2
+            upgrade_light_client_v2(
+                &provider,
+                &mut contracts,
+                is_mock,
+                blocks_per_epoch,
+                epoch_start_block,
+            )
+            .await?;
+        }
+
+        // transfer ownership to multisig
+        let _receipt = transfer_ownership(
+            &provider,
+            Contract::LightClientProxy,
+            lc_proxy_addr,
+            multisig_admin,
+        )
+        .await?;
+        let lc = LightClient::new(lc_proxy_addr, &provider);
+        assert_eq!(lc.owner().call().await?._0, multisig_admin);
+
+        // then send upgrade proposal to the multisig wallet
+        let (result, success) = upgrade_light_client_v2_multisig_owner(
+            &provider,
+            &mut contracts,
+            LightClientV2UpgradeParams {
+                is_mock,
+                blocks_per_epoch,
+                epoch_start_block,
+                rpc_url: sepolia_rpc_url.clone(),
+                dry_run: Some(dry_run),
+            },
+        )
+        .await?;
+        tracing::info!(
+            "Result when trying to upgrade LightClientProxy via the multisig wallet: {:?}",
+            result
+        );
+        assert!(success);
+        if dry_run {
+            let data: serde_json::Value = serde_json::from_str(&result)?;
+            assert_eq!(data["rpcUrl"], sepolia_rpc_url);
+            assert_eq!(data["safeAddress"], multisig_admin.to_string());
+            assert_eq!(data["proxyAddress"], lc_proxy_addr.to_string());
+
+            let expected_init_data = if upgrade_twice {
+                "0x" // no init data for the second upgrade because the proxy was already initialized
+            } else {
+                &LightClientV2::new(lc_proxy_addr, &provider)
+                    .initializeV2(blocks_per_epoch, epoch_start_block)
+                    .calldata()
+                    .to_owned()
+                    .to_string()
+            };
+
+            assert_eq!(data["initData"], expected_init_data.to_string());
+            assert_eq!(data["useHardwareWallet"], false);
+        }
+        // v1 state persistence cannot be tested here because the upgrade proposal is not yet executed
+        // One has to test that the upgrade proposal is available via the Safe UI
+        // and then test that the v1 state is persisted
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_upgrade_light_client_to_v2_multisig_owner_dry_run() -> Result<()> {
+        test_upgrade_light_client_to_v2_multisig_owner_helper(false, true, false).await
+    }
+
+    // We expect no init data for the second upgrade because the proxy was already initialized
+    #[tokio::test]
+    async fn test_upgrade_light_client_to_v2_twice_multisig_owner_dry_run() -> Result<()> {
+        test_upgrade_light_client_to_v2_multisig_owner_helper(false, true, true).await
```

Comment:
The downside of using booleans as arguments is that when reading this code here it's hard to tell what the different values mean. This can be avoided by defining enums instead. Then we also get compile errors if the order is messed up.

Reply:
thank you 
https://github.com/EspressoSystems/espresso-network/commit/e34ebf240cb255e91e01194a1b87ab1c81e25637 & https://github.com/EspressoSystems/espresso-network/commit/95c4aea35c00a49fcee0ca6d402e099f503a8284